File: src/types/character.ts
import { z } from "zod";
import { worldIdSchema } from "./world";
import { eventIdSchema } from "./ids";

/**
 * 角色（Character）— 世界中的人物。
 *
 * A character is NOT a single fixed state — it contains multiple **phases**,
 * each representing a distinct period in the character's life. Scenes reference
 * a character at a specific phase, not the character as a whole.
 */

// ─── Branded IDs ──────────────────────────────────────────────────────────

export const characterIdSchema = z.string().brand<"CharacterId">();
export type CharacterId = z.infer<typeof characterIdSchema>;

export const phaseIdSchema = z.string().brand<"PhaseId">();
export type PhaseId = z.infer<typeof phaseIdSchema>;

// ─── Character Phase ──────────────────────────────────────────────────────

/**
 * A single phase / period in a character's life.
 *
 * Keeps the essential identity anchors (`appearance`, temporal `triggerEventId`)
 * and collapses all narrative-state detail (identity, personality, relationships,
 * abilities, etc.) into a single free-form `changes` field.
 *
 * Phase ordering within a character is determined by array position in
 * `Character.phases` (position 0 = earliest period). No separate `order` field.
 */
export const characterPhaseSchema = z.object({
  id: phaseIdSchema,
  characterId: characterIdSchema,
  /** 外观 — physical appearance description in this period. */
  appearance: z.string(),
  /** 变化 — free-form description of this period's state and changes
   *  (identity, personality, relationships, abilities, etc.). */
  changes: z.string(),
  /** ID of the Event that triggered the transition INTO this phase. `null` for the initial phase. */
  triggerEventId: eventIdSchema.nullable(),
  createdAt: z.iso.datetime(),
  updatedAt: z.iso.datetime(),
});

export type CharacterPhase = z.infer<typeof characterPhaseSchema>;

// ─── Character ────────────────────────────────────────────────────────────

export const characterSchema = z.object({
  id: characterIdSchema,
  worldId: worldIdSchema,
  name: z.string(),
  /** Alternative names, nicknames, or aliases. */
  aliases: z.array(z.string()),
  /** Base-level info that doesn't change across phases. */
  description: z.string(),
  /** Ordered list of phases (position 0 = earliest period). At least one phase always exists. */
  phases: z.array(characterPhaseSchema),
  /** Free-form notes (markdown). */
  notes: z.string(),
  /** User-defined tags for categorization / filtering. */
  tags: z.array(z.string()),
  createdAt: z.iso.datetime(),
  updatedAt: z.iso.datetime(),
});

export type Character = z.infer<typeof characterSchema>;

// ─── Character Reference ──────────────────────────────────────────────────

/**
 * A reference to a character at a specific phase.
 *
 * Used by Scene, Event, and other entities to point at a character in a
 * particular narrative state. Both IDs are foreign keys into the Character /
 * CharacterPhase tables.
 *
 * **Semantics**:
 * - Treated as a composite key — collections of `CharacterRef` (e.g.
 *   `Scene.characterRefs`, `Event.characterRefs`) MUST NOT contain duplicates
 *   (same `characterId` + `phaseId` pair). Enforce at the application/store layer.
 * - **Cascade on delete**: if a `CharacterPhase` is deleted, all refs pointing
 *   at its `phaseId` must be removed (or rewritten to the character's current
 *   phase) by the backend. If a `Character` is deleted, all refs with its
 *   `characterId` are removed. Frontend assumes referential integrity.
 */
export const characterRefSchema = z.object({
  characterId: characterIdSchema,
  phaseId: phaseIdSchema,
});

export type CharacterRef = z.infer<typeof characterRefSchema>;

File: src-tauri/src/models/character.rs
use serde::{Deserialize, Serialize};

/// Reference to a character at a specific phase. Composite key — no duplicates.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CharacterRef {
    pub character_id: String,
    pub phase_id: String,
}

/// A single period in a character's life.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CharacterPhase {
    pub id: String,
    pub character_id: String,
    pub appearance: String,
    pub changes: String,
    pub trigger_event_id: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

/// Input for creating a new phase.
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreatePhaseInput {
    pub appearance: String,
    #[serde(default)]
    pub changes: String,
    #[serde(default)]
    pub trigger_event_id: Option<String>,
}

/// Input for updating a phase (full replacement).
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdatePhaseInput {
    pub appearance: String,
    pub changes: String,
    pub trigger_event_id: Option<String>,
}

/// Character entity with all phases.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Character {
    pub id: String,
    pub world_id: String,
    pub name: String,
    pub aliases: Vec<String>,
    pub description: String,
    pub phases: Vec<CharacterPhase>,
    pub notes: String,
    pub tags: Vec<String>,
    pub created_at: String,
    pub updated_at: String,
}

/// Input for creating a character (requires an initial phase).
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateCharacterInput {
    pub name: String,
    #[serde(default)]
    pub aliases: Vec<String>,
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub notes: String,
    #[serde(default)]
    pub tags: Vec<String>,
    pub initial_phase: CreatePhaseInput,
}

/// Input for updating a character (full replacement, phases managed separately).
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateCharacterInput {
    pub name: String,
    pub aliases: Vec<String>,
    pub description: String,
    pub notes: String,
    pub tags: Vec<String>,
}
